src/plugins/playlist/playlist.js

```diff
@@ -0,0 +1,128 @@
+  const _chainTarget = sliceProperties(options, 'chainTarget').chainTarget;
+  let _playlist = null;
+  let _playlistDisposer = null;
+  let _playlistWidget = null;
+
+  const disposePlaylist = () => {
+    player.removeClass('vjs-playlist');
+    const playlist = player.playlist();
+    _playlist = undefined;
+    playlist.dispose();
+    player.off('cldsourcechanged', _playlistDisposer);
+  };
+
+  const createPlaylist = (sources, options) => {
+    if (sources instanceof Playlist) {
+      _playlist = sources;
+      _playlist.resetState();
+      _playlist.currentIndex(_playlist.currentIndex());
+    } else {
+      _playlist = new Playlist(player.cloudinary, sources, options);
+      _playlist.currentIndex(0);
+    }
+
+    initPlaylistWidget();
+
+    _playlistDisposer = addSourceChangedListener();
+    player.addClass('vjs-playlist');
+  };
+
+  const addSourceChangedListener = () => {
+    const disposer = async () => {
+      if (
+        _playlist &&
+        !_playlist.currentSource().contains(player.currentSource())
+      ) {
+        player.disposePlaylist();
+      }
+    };
+
+    player.on('cldsourcechanged', disposer);
+
+    return disposer;
+  };
+
+  const initPlaylistWidget = () => {
+    player.on(PLAYER_EVENT.PLAYLIST_CREATED, () => {
+      if (_playlistWidget) {
+        _playlistWidget.dispose();
+      }
+
+      if (isPlainObject(options.playlistWidget)) {
+        if (player.fluid_) {
+          options.playlistWidget.fluid = true;
+        }
+
+        if (player.cloudinary.fontFace) {
+          options.playlistWidget.fontFace = player.cloudinary.fontFace;
+        }
+
+        _playlistWidget = new PlaylistWidget(player, options.playlistWidget);
+      }
+    });
+  };
+
+  player.cloudinary.sourcesByTag = (tag, options = {}) => {
+    const url = getCloudinaryUrl(
+      tag,
+      extendCloudinaryConfig(player.cloudinary.cloudinaryConfig(), LIST_BY_TAG_PARAMS)
+    );
+
+    return fetch(url)
+      .then(result => result.json())
+      .then(json => {
+        const resources = normalizeJsonResponse(json.resources);
+
+        if (options.sorter) {
+          resources.sort(options.sorter);
+        }
+
+        const sources = resources.map(resource => {
+          let sourceParams = options.sourceParams || {};
+
+          if (typeof sourceParams === 'function') {
+            sourceParams = sourceParams(resource);
+          }
+
+          const info = (resource.context && resource.context.custom) || {};
+
+          const source = assign({ info }, sourceParams, { publicId: resource.publicId });
+
+          return player.cloudinary.buildSource(source);
+        });
+
+        return sources;
+      });
+  };
+
+  return (sources, options = {}) => {
+    if (sources === undefined) {
+      return _playlist;
+    }
+
+    if (_playlist) {
+      disposePlaylist();
+    }
+
+    createPlaylist(sources, options);
+    player.trigger('playlistcreated');
+
+    return _chainTarget;
+  };
+};
```

Comment:
Frankly the code here looks to be very imperative, this could easily be because the videojs api itself is very imperative. Regardless give it some thought to see if it can be made more declarative.

Reply:
How would you write that? there are many (nicer) patterns we can use here.
Honestly, as much as I agree, I wouldn't do such changes in this phase.
The goal is to separate the playlist plugin because it's big and we believe it's not being used much, not because we want to improve it much :)
Anyway, this patterns are being used throughout the player code-base